keps/sig-node/20190129-hugepages.md

```diff
@@ -304,24 +313,43 @@ container cgroup sandbox will be configured with the specified limit.
 The `kubelet` will ensure the `hugetlb` has no usage charged to the pod level
 cgroup sandbox prior to deleting the pod to ensure all resources are reclaimed.
 
-#### Limits and Quota
+##### Limits and Quota
 
 The `ResourceQuota` resource will be extended to support accounting for
 `hugepages-<hugepagesize>` similar to `cpu` and `memory`.  The `LimitRange`
 resource will be extended to define min and max constraints for `hugepages`
 similar to `cpu` and `memory`.
 
-#### Scheduler changes
+##### Scheduler changes
 
 The scheduler will need to ensure any huge page request defined in the pod spec
 can be fulfilled by a candidate node.
 
-#### cAdvisor changes
+#### cAdvisor changes(Phase 1)
 
 cAdvisor will need to be modified to return the number of pre-allocated huge
 pages per page size on the node.  It will be used to determine capacity and
 calculate allocatable values on the node.
 
+### Phase 2
+
+#### Support container isolation of huge pages
+
+Container isolation of huge pages should be supported to avoid competition between containers to consume huge pages. Currently, `kubelet` sets the agregated huge pages limits on pod's cgroup of hugetlb subsystem. This should be enhanced to set limits on container's cgroup.
```

Comment:
What kind of cgroup are you going to use? Current memory cgroup doesn't have support of limiting NUMA nodes.

Reply:
Container isolation of hugepages does not mean limiting NUMA nodes; it's a related issue but totally different.

At this point, Kubelet supports hugepages in pod isolation, not container isolation.
This is mentioned on an official website([see future section](https://kubernetes.io/docs/tasks/manage-hugepages/scheduling-hugepages/)).
And [this issue](https://github.com/kubernetes/kubernetes/issues/80716) provides more details.

Supporting Pod isolation of hugepage means that the below situation that can happen.
1. The pod has two containers
    (container A and B).
2. Each of the containers requests 2 of 1GB-Hugepages as a resource.
   (container A: 2GB / container B : 2GB)
3. But Kubelet set a limitation as 4GB on pod's cgroup of hugetlb subsystem.
4. It means a single container in a pod can consume 4GB maximally.
5. If a container consumes 2GB maximally, there will be no issue.
    But if a container consumes 4GB, another one cannot consume any hugepage.

Reply:
What kind of cgroup are you going to use?
=> To limiting NUMA nodes, I am going to use cpuset subsystem.
But this KEP and updates will not cover it, [Memory Manager KEP](https://github.com/kubernetes/enhancements/pull/1203) will cover it.